controller/watch/watcher.go

```diff
@@ -249,48 +249,65 @@ func (w *Watcher) Watch(t resourceType, ns string, wcb watchCallback) error {
 }
 
 func deploymentWatcher(di deployments.DeploymentInterface, wcb watchCallback) {
-	for {
-		log.Println("List all existing Deployments")
-		// First do List on the resource to bring things up to date.
-		l, err := di.List(api.ListOptions{})
-		for _, d := range l.Items {
-			log.Printf("Found Deployment name: %s\n", d.Name)
-			event := watch.Event{
-				Type:   watch.Added,
-				Object: &d,
+	// 'listChan' is our signal to know if we should re-get the full list
+	// of resources.
+	listChan := make(chan bool, 1)
+
+	go func() {
+		for <-listChan {
+			log.Println("List all existing Deployments")
+			// Do List on the resource to bring things up to date.
+			l, err := di.List(api.ListOptions{})
+			if err != nil {
+				log.Printf("Error getting list of resources: %v\n", err)
+				continue
+			}
+			for _, d := range l.Items {
+				log.Printf("Found Deployment name: %s\n", d.Name)
+				event := watch.Event{
+					Type:   watch.Added,
+					Object: &d,
+				}
+				wcb(event)
 			}
-			wcb(event)
 		}
+	}()
 
+	// Loop forever, running the watcher loop
+	for {
 		w, err := di.Watch(api.ListOptions{})
 		if err != nil {
 			log.Printf("Failed to start a watch: %v\n", err)
 			continue
 		}
 		c := w.ResultChan()
 
+		// If the watcher is starting (either for the first time or due
+		// to a restart) force us to get a full listing of resources
+		// to make sure we didn't miss anything. Make sure we do it after
+		// the watch has connected though. Only ask for a list if there
+		// isn't a request already queued up.
+		if len(listChan) == 0 {
+			listChan <- true
```

Comment:
@duglin you're right - I was wrongly assuming this for loop was run in multiple goroutines. Nothing to see here 😁